Use Sylius to manage delivery orders

alexsegura opened the pull request:
This PR is a refactoring of what has been done in 29d967fdbc783374135baeccf023aeda65ad99ed to embed a form via an `<iframe>`, to order a delivery. 

It introduces [Sylius](https://github.com/Sylius/Sylius) to manage orders. 

Until #155 is fixed, we cannot properly extend the Sylius models (for example, to link a customer with `OrderInterface`, or even to have our own table names), so I created 2 temporary tables as a workaround: 
- a table to link a Sylius `OrderInterface` with a customer
- a table to link a Sylius `OrderItemInterface` with a delivery

Please note that at the moment, **Sylius orders only contain deliveries**. 
I have started moving Doctrine annotations to YAML, so we will be able to remove those tables shortly, and migrate data properly. 

The process is as before, with a few improvements.

When a customer places an order, it is creates a Sylius `OrderInterface` with `state = cart`, with one `OrderItemInterface` linked to a `Delivery`.

<img width="1155" alt="capture d ecran 2018-03-15 a 13 32 04" src="https://user-images.githubusercontent.com/1162230/37464698-69aaf9fe-2859-11e8-9dba-41a7a4823f69.png">

Administrators are notified and need to confirm the order.

<img width="1155" alt="capture d ecran 2018-03-15 a 13 32 42" src="https://user-images.githubusercontent.com/1162230/37464773-ab0d0068-2859-11e8-82ff-518d416b09c3.png">

--- 

When the order is confirmed, the customer receives an email with a link to a page where the order can be paid by credit card. 

<img width="1243" alt="capture d ecran 2018-03-15 a 13 49 00" src="https://user-images.githubusercontent.com/1162230/37464810-cda54b76-2859-11e8-9d9b-108e44b0af79.png">

Atala commented on app/AppKernel.php:
```diff
@@ -13,8 +13,11 @@ public function registerBundles()
-            new Sylius\Bundle\TaxationBundle\SyliusTaxationBundle(),
             new Sylius\Bundle\ResourceBundle\SyliusResourceBundle(),
+            new Sylius\Bundle\TaxationBundle\SyliusTaxationBundle(),
+            new Sylius\Bundle\MoneyBundle\SyliusMoneyBundle(),
```
Multi-currency support

alexsegura replied:
Yep, but it is just there because `Sylius\Payment` needs it. 
Later, we will be able to manage currencies properly 🙂

Atala commented on app/DoctrineMigrations/Version20180314160032.php:
```diff
@@ -0,0 +1,39 @@
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'postgresql', 'Migration can only be executed safely on \'postgresql\'.');
+
+        $this->addSql('CREATE TABLE delivery_order (order_id INT NOT NULL, user_id INT NOT NULL, PRIMARY KEY(order_id, user_id))');
```
Nice to separate delivery/eshop order into tables

alexsegura replied:
This is actually a workaround 🙂
In the base Sylius Order component, an Order is not linked to anything. 
However, it totally makes sense to link an Order with a Customer. 

When #155 is fixed, we will be able to extend Sylius models, for example to implement [`CustomerAwareInterface`](https://github.com/Sylius/Sylius/blob/f2042d449c5ae3b68df145910cb42fae6c90d254/src/Sylius/Component/Customer/Model/CustomerAwareInterface.php)

Atala commented on app/config/state_machine.yml:
```diff
@@ -0,0 +1,68 @@
+                from: [new]
+                to: processing
+            complete:
+                from: [new, processing]
```
Maybe match with stripe statuses ? Sthg like `captured` or so

Atala commented on src/AppBundle/Controller/PublicController.php:
```diff
@@ -0,0 +1,123 @@
+class PublicController extends Controller
+{
+    /**
+     * @Route("/d/{number}", name="public_delivery")
```
`uuid` is better maybe?

alexsegura replied:
When an `OrderInterface` goes from `state=cart` to `state=new`, it uses the default [`OrderNumberAssigner`](https://github.com/Sylius/Sylius/blob/f2042d449c5ae3b68df145910cb42fae6c90d254/src/Sylius/Bundle/OrderBundle/NumberAssigner/OrderNumberAssigner.php) class. 
I didn't change anything to the defaults, but for sure it can be overriden, and the `order.number` field is a `VARCHAR(255)`. It actually generates _strings_ like `000000001`.

Atala commented on src/AppBundle/Entity/DeliveryOrder.php:
```diff
@@ -0,0 +1,37 @@
+<?php
```
Why not put it in the `Entity/Delivery` folder?

Atala commented on src/AppBundle/Resources/views/Pdf/delivery.html.twig:
```diff
@@ -0,0 +1,94 @@
+<!DOCTYPE html>
+<html>
```
I think to do clean invoices we will need the settings for admin/coop details (adress, siret, aso.).

Atala commented:
Hey !
Looks nice to me :) I added some comments but didn't do the full review yet.
I will test the process tomorrow on my machine

Atala commented:
First of all contragulations for the big amount of work put into this!
I hit two major bugs in the ordering process:
 - bug "phone number is null" in admin before validating delivery (cf screenshot in slack)
 - ~as a customer I didn't received the second mail ("delivery order confirmed") altough everything seems fine in admin~  there is a problem with delivery order confirmation, after clicking the button "confirmed" I lose the button but dont receive the mail + still asked to confirm. Screenshots : 
<img width="1388" alt="capture d ecran 2018-03-16 a 16 47 44" src="https://user-images.githubusercontent.com/4426585/37530435-ced3e710-2939-11e8-8637-08f4eb72b700.png">
<img width="1388" alt="capture d ecran 2018-03-16 a 16 47 52" src="https://user-images.githubusercontent.com/4426585/37530432-cdeaa942-2939-11e8-8986-1e5da555c84a.png">


Other remarks : 
- Wording : Nous allons vous envoyer un email d'ici quelques minutes. -> Nous allons vous envoyer un email d'ici quelques minutes, dès que la commande est validée.
- Visual feedback in admin when delivery order is confirmed (green alert : "Commande confirmée : à vos vélos!")
- No breadcrumb needed (or make it not clickable) as the user is not supposed to go back in the order process (I think? or does it work?)

I also think the UX for the admin is kind of confusing (what is a delivery? a delivery order? how to access both?), we should think twice about it - but not a problem right now.

alexsegura commented:
> bug "phone number is null" in admin before validating delivery (cf screenshot in slack)

Fixed, I was not storing the phone number at all, my bad. 

> after clicking the button "confirmed" I lose the button but dont receive the mail + still asked to confirm

Fixed, once confirmed the warning alert turns into an info alert

<img width="738" alt="capture d ecran 2018-03-17 a 13 21 43" src="https://user-images.githubusercontent.com/1162230/37555356-68ece188-29e6-11e8-8ee7-80e0a618987f.png">

> I also think the UX for the admin is kind of confusing (what is a delivery? a delivery order? how to access both?), we should think twice about it - but not a problem right now.

Yes, sure, this is the next step. 
Of course the goal is to manage all kind of orders (meals & logistics) the same way, & display them on the same admin screen. 
It should be easy to distinguish between orders using [Single Table Inheritance](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/inheritance-mapping.html#single-table-inheritance).
